federationsender/queue/queue.go

```diff
@@ -68,21 +72,24 @@ func (oqs *OutgoingQueues) SendEvent(
 		"destinations": destinations, "event": ev.EventID(),
 	}).Info("Sending event")
 
-	oqs.queuesMutex.Lock()
-	defer oqs.queuesMutex.Unlock()
 	for _, destination := range destinations {
+		oqs.queuesMutex.RLock()
 		oq := oqs.queues[destination]
+		oqs.queuesMutex.RUnlock()
 		if oq == nil {
 			oq = &destinationQueue{
 				rsProducer:  oqs.rsProducer,
 				origin:      oqs.origin,
 				destination: destination,
 				client:      oqs.client,
+				statistics:  oqs.statistics.ForServer(destination),
 			}
+			oqs.queuesMutex.Lock()
 			oqs.queues[destination] = oq
+			oqs.queuesMutex.Unlock()
 		}
 
-		oq.sendEvent(ev)
+		go oq.sendEvent(ev)
```

Comment:
Doesn't this kind of making queuing pointless if you're just going to fork it off to a goroutine? What guarantees ordering?

Reply:
Additionally, I see no point in doing this in a goroutine, as `sendEvent` just chucks it onto a queue then wakes up the worker.

Reply:
This was mostly just a test, I plan to undo that.